### PR TITLE
LSP support for ProgressHandles

### PR DESCRIPTION
**Pull request by sdedic**
~~Still WIP, opened for comments to polish the implementation. Cancel support is still missing.~~

### CODE REVIEW

**Comment by sdedic** [comment]
Possibly ready to review. Licenses and formatting will be corrected in next commit. 

Please focus on NB API changes - I moved some functionality from `org.netbeans.api.progress.nb` to `org.netbeans.api.progress`: the reason is that some features that are of interest of LSP clients use default action that "only" uses some cookies or API, and is not strictly visual. There's a chance that such action can be (in the future) reflected back to the LSP protocol and client. Short term, this allowed to capture progress from Maven, Gradle and others.

Short overview:
- Default (view) action support APIs moved to `org.netbeans.api.progress`
- The message wrapper in LSP `Server` interprets `workDoneProgressToken` in request messages, so potential ProgressHandles connect back to a token that is issued by the LSP client
- The wrapper handles centrally `workDoneCancel` message by cancelling ProgressHandle on behalf of the client.
- `OperationContext` is present in the default lookup (and is propagated through spawned `RequestProcessor.Task`s), ProgressHandle creation works with that to connect back to the client and issue tokens.
- LSP `InternalHandle` implementation sends notification on start, progress and finish events

I adapted certain NB usages of Progress API to use just the basic API, and removed dependencies on Swing-specific implementation, when possible.

File: extide/gradle/nbproject/project.xml [hunk 1]
```diff
@@ -57,15 +57,7 @@
...
                         <release-version>1</release-version>
-                        <specification-version>1.47.1</specification-version>
-                    </run-dependency>
-                </dependency>
-                <dependency>
-                    <code-name-base>org.netbeans.api.progress.nb</code-name-base>
-                    <build-prerequisite/>
-                    <compile-dependency/>
-                    <run-dependency>
-                        <specification-version>1.47.1</specification-version>
```
Comment (JaroslavTulach) [inline extide/gradle/nbproject/project.xml:68]: Removing dependency on `org.netbeans.api.progress.nb` is great!

File: extide/gradle/src/org/netbeans/modules/gradle/api/execute/GradleDistributionManager.java [hunk 2]
```diff
@@ -500,7 +499,7 @@ public String toString() {
         })
         public DownloadTask(GradleDistribution dist) {
             this.dist = dist;
-            handle = ProgressHandleFactory.createSystemHandle(Bundle.TIT_Download_Gradle(dist.getVersion()));
+            handle = ProgressHandle.createSystemHandle(Bundle.TIT_Download_Gradle(dist.getVersion()), null);
```
Comment (JaroslavTulach) [inline extide/gradle/src/org/netbeans/modules/gradle/api/execute/GradleDistributionManager.java:502]: I suggest to create a `META-INF/upgrade/*.hint` [Jackpot hint](https://github.com/oracle/graal/blob/master/truffle/src/com.oracle.truffle.api/src/META-INF/upgrade/sourcebuilder.hint) that would recommend this change. It should be straightforward following [this example](https://github.com/oracle/graal/blob/master/truffle/src/com.oracle.truffle.api/src/META-INF/upgrade/sourcebuilder.hint) moreover I believe @jlahoda can help if there are any issues.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/Debugger.java [hunk 3]
```diff
@@ -48,4 +60,41 @@ public static void startDebugger(InputStream in, OutputStream out) {
...
+        public ConsumeWithLookup(MessageConsumer delegate) {
+            this.delegate = delegate;
+        }
+        
+        @Override
+        public void consume(Message message) throws MessageIssueException, JsonRpcException {
+            InstanceContent ic = new InstanceContent();
+            ProxyLookup ll = new ProxyLookup(new AbstractLookup(ic), Lookup.getDefault());
+            // HACK: piggyback on LSP's client.
+            if (topContext == null) {
```
Comment (JaroslavTulach) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/Debugger.java:77]: OK, probably enough for 12.2.1. Would be good to have this recorded in an issue tracker.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui/AbstractProgressEnvironment.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+import org.netbeans.modules.progress.spi.ProgressEnvironment;
+import org.openide.util.Cancellable;
+import org.openide.util.Lookup;
+import org.openide.util.lookup.Lookups;
+
+/**
+ *
+ * @author sdedic
+ */
+public class AbstractProgressEnvironment implements ProgressEnvironment {
```
Comment (JaroslavTulach) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui/AbstractProgressEnvironment.java:38]: Make `abstract`?

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui/AbstractProgressEnvironment.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    private static final String[] MASK_CANCELLABLES = {
+        "org.netbeans.modules.parsing.impl.indexing.RepositoryUpdater$"
+    };
+    
+    private static final Controller NO_CONTEXT_CONTROLLER = new Controller(null);
+    private final Lookup env;
+    
+    private Pattern patternMaskingCancellables;
+    
+    public AbstractProgressEnvironment() {
```
Comment (JaroslavTulach) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui/AbstractProgressEnvironment.java:53]: Make `protected`.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui/AbstractProgressEnvironment.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        this(Lookups.proxy(() -> Lookup.getDefault()));
+    }
+
+    public AbstractProgressEnvironment(Lookup env) {
```
Comment (JaroslavTulach) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui/AbstractProgressEnvironment.java:57]: Make `protected`.

File: platform/api.progress.nb/src/org/netbeans/api/progress/ProgressHandleFactory.java [hunk 7]
```diff
@@ -65,41 +73,62 @@ public static ProgressHandle createHandle(String displayName, Cancellable allowT
...
+     * Create a progress ui handle for a long lasting task. Since {@code 1.59}, this method was
+     * migrated to {@link ProgressHandle basic Progress API}.
      * @param linkOutput an <code>Action</code> instance that links the running task in the progress bar
      *                   to an output of the task. The action is assumed to open the apropriate component with the task's output.
      * @param displayName to be shown in the progress UI
      * @return an instance of {@link org.netbeans.api.progress.ProgressHandle}, initialized but not started.
+     * @deprecated Please use {@link ProgressHandle#createHandle(java.lang.String, org.openide.util.Cancellable, javax.swing.Action)}
      */
+    @Deprecated
     public static ProgressHandle createHandle(String displayName, Action linkOutput) {
```
Comment (JaroslavTulach) [inline platform/api.progress.nb/src/org/netbeans/api/progress/ProgressHandleFactory.java:85]: Please prepare a Jackpot hint that would rewrite to `ProgressHandle.createHandle`.

File: platform/api.progress.nb/src/org/netbeans/api/progress/ProgressHandleFactory.java [hunk 8]
```diff
@@ -119,44 +148,171 @@ public static JLabel createMainLabelComponent(ProgressHandle handle) {
...
      *          triggers cancel of the task.
      * @param linkOutput an <code>Action</code> instance that links the running task in the progress bar
      *                   to an output of the task. The action is assumed to open the apropriate component with the task's output.
      * @param displayName to be shown in the progress UI
      * @return an instance of {@link org.netbeans.api.progress.ProgressHandle}, initialized but not started.
-     *
+     * @deprecated Please use {@link ProgressHandle#createSystemHandle(java.lang.String, org.openide.util.Cancellable, javax.swing.Action)}
+     * @Deprecated
      */
     public static ProgressHandle createSystemHandle(String displayName, Cancellable allowToCancel, Action linkOutput) {
```
Comment (JaroslavTulach) [inline platform/api.progress.nb/src/org/netbeans/api/progress/ProgressHandleFactory.java:196]: Please prepare a Jackpot hint that would rewrite to `ProgressHandle.createSystemHandle`.

File: platform/api.progress.nb/src/org/netbeans/api/progress/ProgressHandleFactory.java [hunk 9]
```diff
@@ -119,44 +148,171 @@ public static JLabel createMainLabelComponent(ProgressHandle handle) {
...
+            ExtractedProgressUIWorker worker = prov.extractProgressWorker(ih);
+            if (worker != null) {
+                return new StrangeExtractor(ih, worker);
+            } else {
+                return null;
+            }
+        }
+    }
+    
+    private static class StrangeExtractor implements ExtractedProgressUIWorker {
```
Comment (JaroslavTulach) [inline platform/api.progress.nb/src/org/netbeans/api/progress/ProgressHandleFactory.java:239]: `StrangeExtractor` is really a strange name! NonUI, Foreign, Generic, Delegating?

File: platform/api.progress.nb/src/org/netbeans/api/progress/aggregate/AggregateProgressFactory.java [hunk 10]
```diff
@@ -47,12 +48,12 @@ private AggregateProgressFactory() {
      *                   to an output of the task. The action is assumed to open the apropriate component with the task's output.
      * @param displayName to be shown in the progress UI
      * @return an instance of <code>ProgressHandle</code>, initialized but not started.
-     *
+     * @deprecated use {@link BasicAggregateProgressFactory#createHandle(java.lang.String, org.netbeans.api.progress.aggregate.ProgressContributor[], org.openide.util.Cancellable, javax.swing.Action)}.
      */
+    @Deprecated
```
Comment (JaroslavTulach) [inline platform/api.progress.nb/src/org/netbeans/api/progress/aggregate/AggregateProgressFactory.java:53]: Another Jackpot hint, please.

File: platform/api.progress.nb/src/org/netbeans/modules/progress/spi/ProgressUIWorkerProvider.java [hunk 11]
```diff
@@ -29,4 +29,13 @@
     public ProgressUIWorkerWithModel getDefaultWorker();
     
     public ExtractedProgressUIWorker getExtractedComponentWorker();
+
+    /**
+     * Provides an extracted worker instance for the given internal handle.
+     * @param handle internal handle for the worker.
+     * @return the worker instance, possibly {@code null}.
+     */
+    public default ExtractedProgressUIWorker extractProgressWorker(InternalHandle handle) {
```
Comment (JaroslavTulach) [inline platform/api.progress.nb/src/org/netbeans/modules/progress/spi/ProgressUIWorkerProvider.java:38]: Missing versioning of this API change.

File: platform/api.progress.nb/test/unit/src/org/netbeans/api/progress/ProgressHandleFactoryTest.java [hunk 12]
```diff
@@ -179,5 +183,60 @@ private void waitForTimerFinish() {
...
+        // should throw an exception:
+        try {
+            JLabel l = ProgressHandleFactory.createMainLabelComponent(handle);
+            fail("Exppected ISE.");
+        } catch (IllegalStateException ex) {
+            // OK
+        }
+    }
+    
+    public class StrangeHandle extends InternalHandle {
```
Comment (JaroslavTulach) [inline platform/api.progress.nb/test/unit/src/org/netbeans/api/progress/ProgressHandleFactoryTest.java:224]: It is OK to use `StrangeXYZ` in tests, I guess.

File: platform/api.progress/src/org/netbeans/api/progress/ProgressHandle.java [hunk 13]
```diff
@@ -77,6 +86,48 @@ public static ProgressHandle createHandle(String displayName, Cancellable allowT
...
+
+    /**
+     * Creates a possibly cancellable handle, <b>not initiated</b> by the user, e.g. for indexing, 
+     * repository or other system-launched tasks. See documentation for {@link #createHandle(java.lang.String, org.openide.util.Cancellable, java.awt.event.ActionListener)}
+     * for more details.
+     * @param displayName to be shown in the progress UI
+     * @param allowToCancel either {@code null}, if the task cannot be cancelled or an {@link Cancellable} instance
+     * that will be called when the user requests cancel.
+     * @param viewAction listener that will be triggered when the progress is 'activated'.
+     * @return
```
Comment (JaroslavTulach) [inline platform/api.progress/src/org/netbeans/api/progress/ProgressHandle.java:122]: Missing `return` description.

File: platform/api.progress/src/org/netbeans/api/progress/ProgressHandle.java [hunk 14]
```diff
@@ -77,6 +86,48 @@ public static ProgressHandle createHandle(String displayName, Cancellable allowT
...
+        ProgressHandle h = DefaultHandleFactory.get().createHandle(displayName, allowToCancel, true);
+        if (viewAction != null) {
+            h.addDefaultAction(viewAction);
+        }
+        return h;
+    }
+
+    /**
+     * Creates a possibly cancellable handle, <b>not initiated</b> by the user, e.g. for indexing, 
+     * repository or other system-launched tasks. See documentation for {@link #createHandle(java.lang.String, org.openide.util.Cancellable, java.awt.event.ActionListener)}
```
Comment (JaroslavTulach) [inline platform/api.progress/src/org/netbeans/api/progress/ProgressHandle.java:116]: Is the `{@link ...}` correct? I am surprised to see `java.awt.event` package...

File: platform/api.progress/src/org/netbeans/api/progress/ProgressHandle.java [hunk 15]
```diff
@@ -240,6 +291,15 @@ public final void setDisplayName(String newDisplayName) {
         internal.requestDisplayNameChange(newDisplayName);
     }
     
+    /**
+     * Adds an action to the progress indicator. 
+     * @param action describes the action
+     * @return true, if actions are supported by the progress implementation.
```
Comment (JaroslavTulach) [inline platform/api.progress/src/org/netbeans/api/progress/ProgressHandle.java:297]: Missing versioning.

File: platform/api.progress/src/org/netbeans/api/progress/aggregate/BasicAggregateProgressFactory.java [hunk 16]
```diff
@@ -53,6 +54,42 @@ public static ProgressContributor createProgressContributor(String trackingId) {
...
+     * @param allowToCancel either null, if the task cannot be cancelled or 
+     *          an instance of {@link org.openide.util.Cancellable} that will be called when user 
+     *          triggers cancel of the task.
+     * @param linkOutput an <code>Action</code> instance that links the running task in the progress bar
+     *                   to an output of the task. The action is assumed to open the apropriate component with the task's output.
+     * @param displayName to be shown in the progress UI
+     * @return an instance of <code>ProgressHandle</code>, initialized but not started.
+     * @since 1.59
+     */
+    public static AggregateProgressHandle createSystemHandle(String displayName, ProgressContributor[] contributors,
```
Comment (JaroslavTulach) [inline platform/api.progress/src/org/netbeans/api/progress/aggregate/BasicAggregateProgressFactory.java:69]: Probably a bit late to change, but: why the factory methods are in `ProgressHandle`, but not in `AggregateProgressHandle`?

**Review by JaroslavTulach — APPROVED** [review]
The reduced [weight of modules](http://wiki.apidesign.org/wiki/HeavyWeight) looks good. The API looks sane (please add `@since` tags, API change list entries, etc.). Please also consider [fixing the use-cases](https://issues.apache.org/jira/browse/NETBEANS-5215) to adhere to reality.

File: extide/gradle/src/org/netbeans/modules/gradle/api/execute/GradleDistributionManager.java (hunk 2 above)
Comment (sdedic) [reply extide/gradle/src/org/netbeans/modules/gradle/api/execute/GradleDistributionManager.java:502]: Tracked as [NETBEANS-5217](https://issues.apache.org/jira/browse/NETBEANS-5217)

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/Debugger.java (hunk 3 above)
Comment (sdedic) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/Debugger.java:77]: Tracked as [NETBEANS-5218](https://issues.apache.org/jira/browse/NETBEANS-5218)

File: platform/api.progress/src/org/netbeans/api/progress/aggregate/BasicAggregateProgressFactory.java (hunk 16 above)
Comment (sdedic) [reply platform/api.progress/src/org/netbeans/api/progress/aggregate/BasicAggregateProgressFactory.java:69]: I could migrate them (add new ones, deprecated old ones), too - now when the API is being touched. Do you think it is beneficial ?

File: platform/api.progress/src/org/netbeans/api/progress/ProgressHandle.java (hunk 14 above)
Comment (sdedic) [reply platform/api.progress/src/org/netbeans/api/progress/ProgressHandle.java:116]: My bad - I prototyped some variants, one of them accepted just `ActionListener`; but it turned out that some Action features (display name, icon, ...) could be also needed in displayed implementation. So I've ended up with `Action` and forgot to update the javadoc.

**Comment by sdedic** [comment]
Consolidated before merging.

**Review by dbalek — APPROVED** [review]
Looks fine for 12.2.1

**Comment by JaroslavTulach** [comment]
Certainly this fix shall not be backported into `12.0-u2`. Right now it has been integrated, so it will become part of `12.3`. However we want to have it available as part of `12.2-u1` unless I am mistaken. A new PR to backport these changes into `release122` branch has to be created.

**Comment by sdedic** [comment]
oops :) thanks for correction.

**Pull request merged**